File: db/quran/tanzil/translations/scripts/verify.ts
/**
 * verify.ts — integrity checks for the mirrored translations.
 *
 *   tsx scripts/verify.ts
 *
 * Confirms every translation in index.json has a downloaded SQL dump, that
 * each is valid UTF-8 with a sane verse count, and that recorded checksums
 * and metadata are consistent. Exits non-zero on any hard problem.
 */

import { createHash } from "node:crypto";
import { existsSync, readFileSync } from "node:fs";
import path from "node:path";

import { INDEX, ROOT, log, readJson, type IndexFile } from "./lib";

const MIN_AYAS = 6000; // every full-Quran translation should be near 6236

async function main(): Promise<number> {
  const index = await readJson<IndexFile>(INDEX);
  const trans = index.translations;
  const errors: string[] = [];
  const warnings: string[] = [];

  log(`index.json: ${trans.length} translations (header count=${index.count})`);
  if (trans.length !== index.count) errors.push("index count field != actual entries");

  const ayaDist = new Map<number | "none", number>();
  const missing: string[] = [];
  const badSha: string[] = [];
  const badEnc: string[] = [];
  const short: string[] = [];

  for (const tr of trans) {
    const p = path.join(ROOT, tr.file.sql);
    if (!existsSync(p)) {
      missing.push(tr.id);
      continue;
    }
    const raw = readFileSync(p);
    try {
      new TextDecoder("utf-8", { fatal: true }).decode(raw);
    } catch (e) {
      badEnc.push(`${tr.id}: ${(e as Error).message}`);
    }
    if (createHash("sha256").update(raw).digest("hex") !== tr.file.sha256) badSha.push(tr.id);

    const ac = tr.ayaCount ?? "none";
    ayaDist.set(ac, (ayaDist.get(ac) ?? 0) + 1);
    if (typeof ac === "number" && ac < MIN_AYAS) short.push(`${tr.id} (${ac})`);

    for (const f of ["language", "name", "translator", "direction", "languageCode"] as const) {
      if (!tr[f]) warnings.push(`${tr.id}: missing ${f}`);
    }
  }

  const present = trans.length - missing.length;
  log(`\nfiles present: ${present}/${trans.length}`);
  const dist = [...ayaDist.entries()].sort((a, b) => {
    if (a[0] === "none") return 1;
    if (b[0] === "none") return -1;
    return (a[0] as number) - (b[0] as number);
  });
  log(`aya-count distribution: ${JSON.stringify(Object.fromEntries(dist))}`);
  if (short.length) warnings.push(`translations with < ${MIN_AYAS} ayas: ${short.join(", ")}`);

  for (const [label, items] of [["MISSING", missing], ["BAD ENCODING", badEnc], ["BAD SHA256", badSha]] as const) {
    if (items.length) errors.push(`${label}: ${JSON.stringify(items)}`);
  }

  // group by language
  const byLang = new Map<string, number>();
  for (const tr of trans) byLang.set(tr.language, (byLang.get(tr.language) ?? 0) + 1);
  log(`\nlanguages: ${byLang.size}`);
  for (const [lang, n] of [...byLang.entries()].sort((a, b) => b[1] - a[1])) log(`  ${lang.padEnd(14)} ${n}`);

  const rtl = new Set(trans.filter((tr) => tr.direction === "rtl").map((tr) => tr.language));
  log(`\nrtl languages: ${[...rtl].sort().join(", ")}`);

  log("\nWARNINGS:" + (warnings.length ? "" : " none"));
  for (const w of warnings) log(`  ! ${w}`);
  log("ERRORS:" + (errors.length ? "" : " none"));
  for (const e of errors) log(`  ✗ ${e}`);

  return errors.length ? 1 : 0;
}

main().then((code) => process.exit(code));

File: db/quran/tanzil/translations/verify.py
#!/usr/bin/env python3
"""
verify.py — integrity checks for the mirrored tanzil translations.

Confirms every translation in index.json has a downloaded SQL dump, that each
dump is valid UTF-8 with a sane verse count, and that the recorded checksums
and metadata are consistent. Exits non-zero on any hard problem.
"""

from __future__ import annotations

import hashlib
import json
import sys
from collections import Counter
from pathlib import Path

ROOT = Path(__file__).resolve().parent
INDEX = ROOT / "index.json"
MIN_AYAS = 6000  # every full-Quran translation should be near 6236


def main() -> int:
    errors, warnings = [], []
    index = json.loads(INDEX.read_text(encoding="utf-8"))
    trans = index["translations"]
    print(f"index.json: {len(trans)} translations (header count={index['count']})")
    if len(trans) != index["count"]:
        errors.append("index count field != actual entries")

    aya_dist = Counter()
    bad_encoding, bad_sha, missing, short = [], [], [], []

    for t in trans:
        tid = t["id"]
        rel = t["file"]["sql"]
        p = ROOT / rel
        if not p.exists():
            missing.append(tid)
            continue
        raw = p.read_bytes()
        # utf-8 check
        try:
            raw.decode("utf-8")
        except UnicodeDecodeError as e:
            bad_encoding.append(f"{tid}: {e}")
        # checksum check
        if hashlib.sha256(raw).hexdigest() != t["file"]["sha256"]:
            bad_sha.append(tid)
        ac = t.get("ayaCount")
        aya_dist[ac] += 1
        if ac is not None and ac < MIN_AYAS:
            short.append(f"{tid} ({ac})")
        # required display fields present
        for f in ("language", "name", "translator", "direction", "languageCode"):
            if not t.get(f):
                warnings.append(f"{tid}: missing {f}")

    print(f"\nfiles present: {len(trans) - len(missing)}/{len(trans)}")
    print(f"aya-count distribution: {dict(sorted(aya_dist.items(), key=lambda kv: (kv[0] is None, kv[0])))}")
    if short:
        warnings.append(f"translations with < {MIN_AYAS} ayas: " + ", ".join(short))

    for label, items in (("MISSING", missing), ("BAD ENCODING", bad_encoding),
                         ("BAD SHA256", bad_sha)):
        if items:
            errors.append(f"{label}: {items}")

    # group by language for a quick human overview
    by_lang = Counter(t["language"] for t in trans)
    print(f"\nlanguages: {len(by_lang)}")
    for lang, n in by_lang.most_common():
        print(f"  {lang:14} {n}")

    rtl = [t["language"] for t in trans if t["direction"] == "rtl"]
    print(f"\nrtl languages: {sorted(set(rtl))}")

    print("\nWARNINGS:" + ("" if not warnings else ""))
    for w in warnings:
        print(f"  ! {w}")
    print("ERRORS:" + ("" if not errors else ""))
    for e in errors:
        print(f"  ✗ {e}")

    return 1 if errors else 0


if __name__ == "__main__":
    sys.exit(main())
